modules/payloads/stages/windows/custom.rb

```diff
@@ -0,0 +1,38 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+module MetasploitModule
+  include Msf::Payload::Windows
+  include Msf::Payload::Custom
+  include Msf::Payload::Custom::Options
+
+  def initialize(info = {})
+    super(
+      merge_info(
+        info,
+        'Name' => 'Windows shellcode stager',
+        'Description' => 'Custom shellcode stager',
+        'Author' => 'bwatters-r7',
+        'License' => MSF_LICENSE,
+        'Platform' => 'win',
+        'Arch' => ARCH_X86,
+        'Session' => nil,
+        'PayloadCompat' => {}
+      )
+    )
+  end
+
+  def stage_payload(_opts = {})
+    print_line('Trying to stage Payload')
+    unless datastore['SHELLCODE_FILE'].nil?
+      shellcode = File.binread(datastore['SHELLCODE_FILE'])
+      if datastore['PrependSize']
+        return [ shellcode.length ].pack('V') + shellcode
```

Comment:
I haven't tested this but won't this result in the size being prepended twice when a TCP stager is used? It looks like this should only be prepended for the reverse_http(s) stagers.

Reply:
I don't believe so.  The prepending of size is done in the method `handle_intermediate_stage`, (https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/core/payload/windows.rb#L100) which is called in `handle_connection` https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/core/payload/stager.rb#L208
To the best of my knowledge, we prepend the size as part of the handler code, so I located the prepended size for custom payloads in the handler as well.  If you staged a second stager, then each handler would send the size, but that would be the expected behavior.
That said, `handle_intermediate_stage` is redefined 26 times and called once, so it is possible I missed something.

Reply:
"That said, handle_intermediate_stage is redefined 26 times and called once" is haunting excerpt, thanks. :wink: